stellargraph/losses.py

```diff
@@ -49,3 +54,49 @@ def graph_log_likelihood(batch_adj, wys_output):
     loss = tf.math.reduce_sum(tf.abs(matrix))
 
     return tf.expand_dims(loss, 0)
+
+
+class SelfAdversarialNegativeSampling(tf.keras.losses.Loss):
+    """
```

Comment:
did you see the implementation of this at?: https://github.com/DeepGraphLearning/KnowledgeGraphEmbedding/blob/2442e8b9bb1a40e44edabcb3cc4f145ac24b1c7b/codes/model.py#L251

Reply:
Yeah, I did briefly look at it. I think their non-Keras approach makes their code simpler for this particular example, because they're having to implement the full training step, so they have more detailed control over how the original sampling connects to the loss function. It also means it doesn't translate that directly into our implementation here.

That said, looking at it a second time, I realised that the implementation here isn't _quite_ right: this PR doesn't do the equivalent of the `detach` https://github.com/DeepGraphLearning/KnowledgeGraphEmbedding/blob/2442e8b9bb1a40e44edabcb3cc4f145ac24b1c7b/codes/model.py#L269-L272

I'll add a call to https://www.tensorflow.org/api_docs/python/tf/stop_gradient which should be equivalent.